### PR TITLE
Migrate grammar-plugin to 5th version of SonarQube

### PR DESCRIPTION
**Pull request by OlegSopilnyak**
Rewritten issues registration approach

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -102,15 +123,25 @@
...
         </plugins>
         <pluginManagement>
             <plugins>
-                <plugin>
-                    <groupId>org.codehaus.sonar</groupId>
-                    <artifactId>sonar-packaging-maven-plugin</artifactId>
-                    <version>1.9</version>
-                </plugin>
-            </plugins>
+<!--                    <plugin>
```
Comment (webdizz) [inline pom.xml]: please remove commented code

File: src/main/java/name/webdizz/sonar/grammar/PluginParameter.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package name.webdizz.sonar.grammar;
+
+import org.sonar.api.CoreProperties;
+
+/**
+ * The pool of constants used in the grammar-plugin
+ *
+ * @author Oleg_Sopilnyak1
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/PluginParameter.java]: this is not necessary, SCM know who did what

File: src/main/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinition.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.slf4j.LoggerFactory;
+import org.sonar.api.BatchExtension;
+import org.sonar.api.rule.RuleStatus;
+import org.sonar.api.rule.Severity;
+import org.sonar.api.server.rule.RulesDefinition;
+
+/**
+ * Declare rule metadata in server repository of rules.
+ *
+ * @author Oleg_Sopilnyak1
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinition.java]: the same as above

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        int lineCounter = 1;
+        try {
+            final List<String> code = FileUtils.readLines(inputFile.file());
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Has read code from {}", inputFile.absolutePath());
+            }
+
+            for (final String line : code) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Line {}:{}", lineCounter, line);
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java]: add a bit more details

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    @Override
+    public void analyse(Project module, SensorContext context) {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Initialize the GrammarChecker.");
+        }
+        grammarChecker.initialize();
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Checking {}-files.", PluginParameter.PROFILE_LANGUAGE);
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java]: add more information about context

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    public void incident(final String message, final int column) {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Reported about incident \"{}\" Col. {}", message, column);
+        }
+        final Issuable issuable = perspectives.as(Issuable.class, inputFile);
+        final Issue issue = issuable.newIssueBuilder()
+                .message(message)
+                .line(new Integer(lineNumber))
+                .ruleKey(ruleKey)
+                .attribute("Col.", Integer.toString(column))
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java]: Should be a constant

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+                .ruleKey(ruleKey)
+                .build();
+        final boolean success = issuable.addIssue(issue);
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Issue about incident has added {}", success);
+        }
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java]: Object methods are better to be placed at the class end as there is no such interest in them

File: src/test/java/name/webdizz/sonar/grammar/sensor/GrammarViolationTriggerTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+public class GrammarViolationTriggerTest {
+    
+    public GrammarViolationTriggerTest() {
+    }
+
+    /**
+     * Test of spellingError method, of class GrammarViolationTrigger.
+     */
+    @Test
+    public void testSpellingError() {
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/sensor/GrammarViolationTriggerTest.java]: too big test, it's not clear what we're testing
please split into several small

Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/sensor/GrammarViolationTriggerTest.java]: BTW test usually starts with should+narrative what is being tested

File: src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import org.sonar.api.profiles.RulesProfile;
+import org.sonar.api.rules.Rule;
+import org.sonar.api.rules.RuleFinder;
+import org.sonar.api.utils.ValidationMessages;
+
+/**
+ *
+ * @author Oleg_Sopilnyak1
+ */
+//@RunWith(PowerMockRunner.class)
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java]: do not use power mock until it's really required

File: src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+//@PrepareForTest(RulesProfile.class)
+public class GrammarProfileDefinitionTest {
+
+    private final RuleFinder ruleFinder = mock(RuleFinder.class);
+
+    /**
+     * Test of createProfile method, of class GrammarProfileDefinition.
+     */
+    @Test
+    public void testCreateProfile() {
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java]: please split into more granular tests methods

File: src/test/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinitionTest.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+public class GrammarRulesDefinitionTest {
+    
+    public GrammarRulesDefinitionTest() {
+    }
+
+    /**
+     * Test of define method, of class GrammarRulesDefinition.
+     */
+    @Test
+    public void testDefine() {
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinitionTest.java]: please apply already mentioned convention for tests

File: pom.xml (hunk 1 above)
Comment (OlegSopilnyak) [inline pom.xml]: Done.

File: src/main/java/name/webdizz/sonar/grammar/PluginParameter.java (hunk 2 above)
Comment (OlegSopilnyak) [inline src/main/java/name/webdizz/sonar/grammar/PluginParameter.java]: Done.

File: src/main/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinition.java (hunk 3 above)
Comment (OlegSopilnyak) [inline src/main/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinition.java]: Done.

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java (hunk 4 above)
Comment (OlegSopilnyak) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java]: Done.

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java (hunk 5 above)
Comment (OlegSopilnyak) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesSensor.java]: Added information about module and context.toString()

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java (hunk 6 above)
Comment (OlegSopilnyak) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java]: Done.

File: src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java (hunk 7 above)
Comment (OlegSopilnyak) [inline src/main/java/name/webdizz/sonar/grammar/sensor/GrammarIssuesWrapper.java]: Removed unnecessary methods (hashCode() & equals(...))

File: src/test/java/name/webdizz/sonar/grammar/sensor/GrammarViolationTriggerTest.java (hunk 8 above)
Comment (OlegSopilnyak) [inline src/test/java/name/webdizz/sonar/grammar/sensor/GrammarViolationTriggerTest.java]: Made refactoring of the class

File: src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java (hunk 9 above)
Comment (OlegSopilnyak) [inline src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java]: Done.

File: src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java (hunk 10 above)
Comment (OlegSopilnyak) [inline src/test/java/name/webdizz/sonar/grammar/profile/GrammarProfileDefinitionTest.java]: Made refactoring of the class

File: src/test/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinitionTest.java (hunk 11 above)
Comment (OlegSopilnyak) [inline src/test/java/name/webdizz/sonar/grammar/rule/GrammarRulesDefinitionTest.java]: Made refactoring of the class

**Pull request merged**